File: src/ProjectR/content/scripts/afflictions/Debuffs/MiniStun.lua
function GetName()
  return "Mini Stun"
end

function GetType()
  return Debuff
end

function GetHookPoints()
  return { TurnCounterUpdating }
end

function OnAttachment(char)
  local parStrength = GetTimeToAction()

  char:SetVar("ministun_Strength", parStrength)
end

function OnRemoval(char)
end

function Trigger(hookPoint, char)
  if hookPoint ~= TurnCounterUpdating then
    return
  end

  local timeStep = GetStats(char):GetTotalStat(SPD)
  local currStrength = char:GetVar("ministun_Strength")
  local newStrength = currStrength - timeStep
  char:SetVar("ministun_Strength", newStrength)

  if newStrength <= 0 then
    this:RemoveFrom(char)
  end

  return false
end

File: src/ProjectR.Scripting/content/scripts/afflictions/Debuffs/MiniStun.cs
using ProjectR.Interfaces.Model;
using ProjectR.Interfaces.Model.Stats;

namespace ProjectR.Scripting.Afflictions
{
    public class MiniStun : Affliction
    {
        public override string Name { get { return "Mini Stun"; } }

        public override AfflictionType Type { get { return AfflictionType.Debuff; } }

        protected override HookPoint[] HookPoints
        {
            get
            {
                return new[]
                {
                    HookPoint.TurnCounterUpdating, 
                };
            }
        }

        protected override void OnAttachment(ICharacter character)
        {
            var parStrength = character.TimeToAction;
            SetVar(character, "ministun_Strength", parStrength);
        }

        protected override void OnTurnCounterUpdating(ICharacter character, BoolConsolidator result)
        {
            var timeStep = character.Stats.GetTotalStat(BaseStat.SPD);
            var currentStrength = GetVar(character, "ministun_Strength");
            var newStrength = currentStrength - timeStep;
            SetVar(character, "ministun_Strength", newStrength);

            if (newStrength <= 0)
            {
                RemoveFrom(character);
            }

            result.AddResult(false);
        }
    }
}